src/dmd/traits.d

```diff
@@ -1416,11 +1416,29 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
         if (dim != 1)
             return dimError(1);
 
+        void usage()
+        {
+            if (e.ident == Id.allMembers)
+            {
+                e.errorSupplemental("In `__traits( allMembers, TraitsArgument )` `TraitsArgument` must evaluate to either a type or an expression of type");
+                e.errorSupplemental("It shouldn't be basic data type, dynamic, static or associative array or not instantiated template");
+                e.errorSupplemental("For more info see https://dlang.org/spec/traits.html#allMembers");
+            }
```

Comment:
```suggestion
            e.errorSupplemental("In `__traits( %s, TraitsArgument )` `TraitsArgument` must evaluate to either a type or an expression of type");
            e.errorSupplemental("It shouldn't be basic data type, dynamic, static or associative array or not instantiated template");
            e.errorSupplemental("For more info see https://dlang.org/spec/traits.html#%s", e.ident, e.ident);
```

And remove the second branch.

However:
- 3 errorSupplemental is a bit much.
- "expression of type" is not the correct wording.
- We could print the expression here, either it's `toChars()` or its `kind()`

I suppose providing an exhaustive list of supported symbols kind would be too long ?

Reply:
Also, can we replace `TraitsArgument` with the exact argument that was passed?

Reply:
> ` e.errorSupplemental("It shouldn't be basic data type, dynamic, static or associative array or not instantiated template");`

Additionally, instead of listing what the argument should not be, it's much better to list what it should be: an aggregate: 
- module
- struct
- class
- interface
- template (non-eponymous templates having multiple members are supported, right?)

Reply:
Also `import` it seems ?

Reply:
If all of these 3 lines are useful then it's ok.

I agree to you about "expression of type". It sounds strange even to me, non native English speaker, but the offical docs [says](https://dlang.org/spec/traits.html#allMembers) so. Here I need a help to make a correct text in English.

Yes, the expression should be printed too like:
```
source/aux/traits.d(132,31): Error: In expression __traits(allMembers, TaggedAlgebraic!(Payload)[]) TaggedAlgebraic!(Payload)[] has no members
...
```
Also I think that `has no members` isn't a right statement, the meaning should be more like `it can't have any members` because `struct Foo {}` has no members, right, but `float` can't have them at all.
So I'll add the expression printing and your changes to `Usage` function but I need a help to construct right English message text.

Reply:
an exhaustive list of supported symbols kind as I understand contains only aggregate types and instantiated(?) templates and expressions that can be resolved to these types. But I'm noly starting learning dmd so can be wrong.

Reply:
Also I can add more details about why it can not have members:
```
source/aux/traits.d(132,31): Error: In expression __traits(allMembers, TaggedAlgebraic!(Payload)[]) TaggedAlgebraic!(Payload)[] can't have members because it is a dynamic array
```
But that means more source code becase RootObject has no `kind()` and one need to cast it to all possible types to detect its kind. But `semanticTraits()` function already is huge, so I don't know.

Reply:
Oh, I didn't refresh the page before anwering...